[MLIR] [Transforms] Let `transform.structured.convert_to_loops` return handles to loops

lhunloh opened the pull request:
This lets `transform.structured.convert_to_loops` return handles to the generated loops, making this transformation more useful to use for (transformation-)nesting purposes. This is modelled after SCFs `transform.loop.forall_to_for` which returns handles to loops.

Introduced in commit aa2a96a24ae3a8cc04635ab6ede474c5f2665053, with a note that they might move out of the `Linalg`-Dialect, but no reason given for the non-return of handles. As far as I can see, this transform always returns loops.

lhunloh commented:
Pinging @MaheshRavishankar as I don't have permission to add reviewers and it's your initial code. :)

qedawkins commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2122,6 +2122,9 @@ DiagnosedSilenceableFailure transform::ConvertToLoopsOp::applyToOne(
   rewriter.eraseOp(target);
+  for (auto &loop : *loops) {
+    results.push_back(loop);
+  }
```
I'm not sure this is assigning the loops to the results correctly. This is appending to a `transform::ApplyToEachResultList` for each payload operation, which then gets mapped 1:1 with the results of the transform op. This will only work if all converted linalg ops produce the same number of loops. This seems undesirable to me in cases where users just want to convert all linalg ops to loops, or can't guarantee they know the rank of the linalg op they are converting. I see 2 options

1. Remove the `TransformEachOpTrait` and collect all of the converted loops into a single handle.
2. Remove the `TransformEachOpTrait` and ignore the loop handles if `convert_to_loops` has no results.

https://github.com/llvm/llvm-project/blob/a64975f9660e02f5f6688f8bcc55daf9eaa99fda/mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp#L1462
and where it checks the exact number of results: https://github.com/llvm/llvm-project/blob/a64975f9660e02f5f6688f8bcc55daf9eaa99fda/mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp#L1511

lhunloh replied:
We decided to go with a single handle for all generated loops. This is easily splittable via `transform.split_handle`.

ftynse commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -2122,6 +2122,9 @@ DiagnosedSilenceableFailure transform::ConvertToLoopsOp::applyToOne(
   if (failed(loops))
     return emitDefaultDefiniteFailure(target);
   rewriter.eraseOp(target);
+  for (auto &loop : *loops) {
```
Nit: please expand `auto` unless the type is obvious from line-level context.

ftynse reviewed (commented):
Could we have a test for this?

lhunloh commented:
Test added, review feedback worked in. :)
We now return a single handle to all generated loops, which is easily splittable via `transform.split_handle`.

qedawkins commented on mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:
```diff
@@ -1274,33 +1274,28 @@ def ScalarizeOp : Op<Transform_Dialect, "structured.scalarize",
     For operations that implement the `TilingInterface`, and implement
     the `generateScalarImplementation` method, lowers the operation to
-    loops. This operation does not return any handles.
+    loops. The return handle points to all generated loops.
```
nit: Add to the op description that it emits a definite failure if the pattern to lower the payload operation to loops fails.

lhunloh replied:
Added "Fails if the payload ops cannot be lowered to loops."

qedawkins reviewed (approved):
nice, thanks!

Commit:
Let `transform.structured.convert_to_loops` return handle to loops

This lets `transform.structured.convert_to_loops` return a handle to all
generated loops, making this transformation more useful to use
for (transformation-)nesting purposes. This handle may be split via `transform.split_handle` to get handles to each individual loop.

lhunloh commented:
Commits are squashed, I don't have write access to merge, if everythings looks good now. :) @ftynse @qedawkins
